### PR TITLE
gh-98778: Update HTTPError to initialize properly even if fp is None

### PR DESCRIPTION
**Pull request by corona10**
* Issue: gh-98778

### CODE REVIEW

**Comment by corona10** [comment]
Since the `_TemporaryFileCloser` is used for internal usage, this change will not impact to 3rd party libraries.

**Comment by corona10** [comment]
````

(.oss) ➜  cpython git:(gh-98778) ✗ ./python.exe -m test test_urllib2 -R 3:3
Raised RLIMIT_NOFILE: 256 -> 1024
0:00:00 load avg: 1.44 Run tests sequentially
0:00:00 load avg: 1.44 [1/1] test_urllib2
beginning 6 repetitions
123456
......

== Tests result: SUCCESS ==

1 test OK.

Total duration: 1.5 sec
Tests result: SUCCESS
````

**Review by ambv — CHANGES_REQUESTED** [review]
You made changes to `_TemporaryFileCloser` but `urllib.response.addinfourl` here is a subclass of `_TemporaryFileWrapper`. Now after your changes the `.close()` and `.cleanup()` operations work but the wrapper is still in an inconsistent state. It cannot be used as a context manager, and cannot be iterated. The wrapper proxies attribute access to the file handle so asking things like `exc.mode`, `exc.seekable()`, etc. will all fail.

The cleanest solution would be to change `HTTPError` to not subclass this thing at all... but that ship has sailed a long time ago.

Your change weakens `_TemporaryFileCloser` by allowing it to silently accept `None` in lieu of a file handle. This makes `tempfile` worse, I'm not convinced this is a good compromise.

Instead, since `HTTPError` is already a hack (and there was literally a comment about why it's hacky where the bug was created), I would just replace the hack with a different one: set `fp` to `io.StringIO()` if the passed `fp` is `None`. It's still not perfect but I feel that would be more robust.

**Comment by corona10** [comment]
> Instead, since HTTPError is already a hack (and there was literally a comment about why it's hacky where the bug was created), I would just replace the hack with a different one: set fp to io.StringIO() if the passed fp is None. It's still not perfect but I feel that would be more robust.

+1

**Comment by corona10** [comment]
@ambv 

Thank you for the review and for pointing out what I missed :)

I have made the requested changes; please review again

File: Lib/urllib/error.py [hunk 1]
```diff
@@ -42,12 +42,10 @@ def __init__(self, url, code, msg, hdrs, fp):
...
         self.fp = fp
         self.filename = url
-        # The addinfourl classes depend on fp being a valid file
-        # object.  In some cases, the HTTPError may not have a valid
-        # file object.  If this happens, the simplest workaround is to
-        # not initialize the base classes.
         if fp is not None:
             self.__super_init(fp, hdrs, url, code)
+        else:
+            self.__super_init(io.StringIO(), hdrs, url, code)
```
Comment (ambv) [inline Lib/urllib/error.py]: It's both fewer lines and less redundant to do it in one go.

```suggestion
        if fp is None:
            fp = io.StringIO()
        self.__super_init(fp, hdrs, url, code)
```

**Review by ambv — APPROVED** [review]

File: Lib/urllib/error.py [hunk 2]
```diff
@@ -42,12 +42,9 @@ def __init__(self, url, code, msg, hdrs, fp):
...
         self.fp = fp
         self.filename = url
-        # The addinfourl classes depend on fp being a valid file
-        # object.  In some cases, the HTTPError may not have a valid
-        # file object.  If this happens, the simplest workaround is to
-        # not initialize the base classes.
-        if fp is not None:
-            self.__super_init(fp, hdrs, url, code)
+        if fp is None:
+            fp = io.StringIO()
```
Comment (gaborbernat) [inline Lib/urllib/error.py:46]: Shouldn't this be BytesIO? 

```python
from urllib.request import urlopen
from urllib.error import HTTPError
try:
    urlopen('http://asadsad.sd')
except HTTPError as exception:
    content = exception.fp.read()
    print(type(content))
```
```
<class 'bytes'>
```

**Pull request merged**